// koid/dist/lib/helper.js
export function parseConfig(config) {
    /* istanbul ignore else */
    if (typeof config === 'undefined') {
        const dc = genConfigRandom();
        return parseConfigDc(dc);
    }
    if (typeof config.node === 'number') {
        const configNode = {
            node: config.node,
            epoch: config.epoch,
        };
        return parseConfigNode(configNode);
    }
    else {
        const conf = config;
        const configDc = {
            dataCenter: conf.dataCenter,
            worker: conf.worker,
            epoch: conf.epoch,
        };
        return parseConfigDc(configDc);
    }
}
/**
 * Generate random id
 */
export function genConfigRandom() {
    const id = Date.now() & 0x3FF;
    const config = parseConfigNode({ node: id });
    const ret = {
        dataCenter: config.dataCenter,
        worker: config.worker,
        epoch: 0,
    };
    return ret;
}
function parseConfigDc(config) {
    const dataCenter = config.dataCenter & 0x1F;
    const worker = config.worker & 0x1F;
    const id = (dataCenter << 5 | worker) & 0x3FF;
    const opts = {
        genId: id << 12,
        epoch: typeof config.epoch === 'number' ? config.epoch : 0,
        dataCenter,
        worker,
    };
    return opts;
}
function parseConfigNode(config) {
    const id = config.node & 0x3FF;
    const dataCenter = id >> 5;
    const worker = id & 0x1F;
    const opts = {
        genId: id << 12,
        epoch: typeof config.epoch === 'number' ? config.epoch : 0,
        dataCenter,
        worker,
    };
    return opts;
}
